### PR TITLE
Create pipeline to train Random Forest models from an atlas 

### PR DESCRIPTION
**Pull request by gboggy**

### CODE REVIEW

**Comment by GWMcElfresh** [comment]
Hi all, I added hyperparameter tuning, XGBoost functionality, and operations on a seurat object into Pipeline_Seurat.R 

It's effectively a translation of Pipeline.R to work directly on a seurat object. The `TrainAllModels()` function requires a seuratObj with a ground truth celltype column. I'll put a usage .Rmd in our teams chat. 

Critically, I dropped support for subsetting the training_matrix using a list of genes. I think that doing feature subsetting is convenient while we're training models on our laptops, but rather inconvenient for missing genes that we don't know about _a priori_ and we may discover via UCell smoothing or similar.

I think that these models will still classify on matrices using a subset of features but again, no need to sacrifice prediction power when the ultimate goal is to train these models once per RIRA edition.

I would advise for the time being that we implement new visualizations/methods in the more modular Pipeline.R that we can easily run locally and then migrate the ones we decide to keep into Pipeline_Seurat.R and throw it onto a cluster node or five for final training.

**Review by bbimber — CHANGES_REQUESTED** [review]
@GWMcElfresh (or anyone), some general comments:
- I started a test case that tries to use seurat's PBMC3k as input. The purpose of this isnt necessarily to exercise real calling, but to exercise the code from start to finish on a dataset that is small enough to run fast-ish, but big enough to make it close enough to real. I dont know where the right balance is, but I doubt the first cut is there. 
- What about serializing one RDS with all the classifiers as a list to disk, rather than per-celltype RDS files? The reasoning is that our cell types might not be legal filenames, and as-is we dont have a way to track those identifiers except for in the filename itself. I suppose we could also serialize a list with list(celltype=XXX, classifier=XXXX) and still keep it per-cell type?
- Do we need to be careful about renaming features from '.' to '-'? This is b/c seurat doesnt like hyphens and I guess we need to remove dots for the RF code? Could we end up with gene names that dont translate back and forth easily?

**Comment by bbimber** [comment]
Another thing, do you expect DALEX::model_parts to take forever to run?
```
  parts <- DALEX::model_parts(explainer, type = plot_type)
```

**Comment by GWMcElfresh** [comment]
> @GWMcElfresh (or anyone), some general comments:
> 
> * I started a test case that tries to use seurat's PBMC3k as input. The purpose of this isnt necessarily to exercise real calling, but to exercise the code from start to finish on a dataset that is small enough to run fast-ish, but big enough to make it close enough to real. I dont know where the right balance is, but I doubt the first cut is there.

I think PBMC3k is the right move. The resampling techniques in the hyperparameter training don't seem to work very well on `pbmc_small`, but we could think about cutting down on the feature space to make it run faster. 

> * What about serializing one RDS with all the classifiers as a list to disk, rather than per-celltype RDS files? The reasoning is that our cell types might not be legal filenames, and as-is we dont have a way to track those identifiers except for in the filename itself. I suppose we could also serialize a list with list(celltype=XXX, classifier=XXXX) and still keep it per-cell type?

This seems fine. I suppose the main concern is passing greek symbols? It's vaguely beneficial to pull down individual models, but the difference is marginal enough to be useful either way. 

> * Do we need to be careful about renaming features from '.' to '-'? This is b/c seurat doesnt like hyphens and I guess we need to remove dots for the RF code? Could we end up with gene names that dont translate back and forth easily?

This should be okay. We run into errors with genes with the same name in Mmul10 that need to be resolved with `make.unique()` which by default suffixes with '.#' In the event that we need to swap away from '.' I think it makes most sense to do a `gsub`  after all of the classification steps, since we'll likely introduce periods in `make.unique`.

> * Another thing, do you expect DALEX::model_parts to take forever to run?

Yeah. `model_parts` does a more proper permutation for feature importance, but it has cubic scaling with the number of genes. 
It is possible to tune it to run faster. The number of permutations `B` parameter is default 10 and the number of observations (for accuracy measurement) parameter `N` is default 1000, both of which might decrease computation time. 
I would suggest leaving these (and perhaps increasing B) and using `model_parts` as both a more rigorous way to check simple/exploratory models with imbalanced training sets and a final finishing step for models that would be coupled with a RIRA atlas edition. 

I could/should extend this function to calculate Shapley values as well.
